src/EditorFeatures/Core/Implementation/IntelliSense/AsyncCompletion/ItemManager.cs

```diff
@@ -0,0 +1,477 @@
+
+using System;
+using System.Collections.Generic;
+using System.Collections.Immutable;
+using System.Globalization;
+using System.Linq;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.CodeAnalysis;
+using Microsoft.CodeAnalysis.Completion;
+using Microsoft.CodeAnalysis.Editor.Shared.Utilities;
+using Microsoft.CodeAnalysis.Shared.Extensions;
+using Microsoft.CodeAnalysis.Text;
+using Microsoft.CodeAnalysis.Text.Shared.Extensions;
+using Microsoft.VisualStudio.Language.Intellisense.AsyncCompletion;
+using Microsoft.VisualStudio.Text;
+using Roslyn.Utilities;
+using AsyncCompletionData = Microsoft.VisualStudio.Language.Intellisense.AsyncCompletion.Data;
+using FilterResult = Microsoft.CodeAnalysis.Editor.Implementation.IntelliSense.Completion.Controller.Session.FilterResult;
+using RoslynCompletionItem = Microsoft.CodeAnalysis.Completion.CompletionItem;
+using Session = Microsoft.CodeAnalysis.Editor.Implementation.IntelliSense.Completion.Controller.Session;
+using VSCompletionItem = Microsoft.VisualStudio.Language.Intellisense.AsyncCompletion.Data.CompletionItem;
+
+namespace Microsoft.CodeAnalysis.Editor.Implementation.IntelliSense.Completion.AsyncCompletion
+{
+    internal class ItemManager : ForegroundThreadAffinitizedObject, IAsyncCompletionItemManager
+    {
+        private readonly CompletionHelper _completionHelper;
+
+        private const int MaxMRUSize = 10;
+        private ImmutableArray<string> _recentItems = ImmutableArray<string>.Empty;
+
+        internal ItemManager(IThreadingContext threadingContext) : base(threadingContext)
+        {
+            _completionHelper = new CompletionHelper(isCaseSensitive: true);
+        }
+
+        public Task<ImmutableArray<VSCompletionItem>> SortCompletionListAsync(
+            IAsyncCompletionSession session,
+            AsyncCompletionData.AsyncCompletionSessionInitialDataSnapshot data,
+            CancellationToken cancellationToken)
+        {
+            AssertIsBackground();
+
+            SubscribeEvents(session);
+            return Task.FromResult(data.InitialList.OrderBy(i => i.SortText).ToImmutableArray());
+        }
+
+        public Task<AsyncCompletionData.FilteredCompletionModel> UpdateCompletionListAsync(
+            IAsyncCompletionSession session,
+            AsyncCompletionData.AsyncCompletionSessionDataSnapshot data,
+            CancellationToken cancellationToken)
+            => Task.FromResult(UpdateCompletionList(session, data, cancellationToken));
+
+        private AsyncCompletionData.FilteredCompletionModel UpdateCompletionList(
+            IAsyncCompletionSession session,
+            AsyncCompletionData.AsyncCompletionSessionDataSnapshot data,
+            CancellationToken cancellationToken)
+        {
+            AssertIsBackground();
+
+            SubscribeEvents(session);
```

Comment:
what is this for?

Reply:
Copied from above:
1. We'd like to update the list of most recent items on SessionCommitted. Here and in the update method we're subscribing for this event.
2. We'd like to unsubscribe from this event when session is dismissed. For the purpose, we also subscribe for SessionDismissed.

---
In reply to: [232509521](https://github.com/dotnet/roslyn/pull/29016#discussion_r232509521) [](ancestors = 232509521)

Reply:
sorry, i guess i'm not being clear.  it seems like such a strane place to subscribe to events.  and it looked like tehre were multiple paths toward subscription.

Reply:
> Can we make it better?

Can we not just listen for the session being created and subscribe there?

Reply:
Let us try.

---
In reply to: [233250742](https://github.com/dotnet/roslyn/pull/29016#discussion_r233250742) [](ancestors = 233250742)

Reply:
Removed the events.

---
In reply to: [233247540](https://github.com/dotnet/roslyn/pull/29016#discussion_r233247540) [](ancestors = 233247540,233242623)